File: backend/functions/app/services/chat_history_service.py
"""
backend/functions/app/services/chat_history_service.py
Chat history management service for storing and retrieving conversations
"""

import firebase_admin
from firebase_admin import firestore
from datetime import datetime, timedelta
import logging
from typing import List, Dict, Any, Optional
from collections import Counter

logger = logging.getLogger(__name__)


class ChatHistoryService:
    """Service for managing chat history"""
    
    def __init__(self):
        """Initialize chat history service"""
        try:
            self.db = firestore.client()
            self.history_collection = 'chat_history'
            self.message_retention_days = 90  # Default retention period
            logger.info("ChatHistoryService initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize ChatHistoryService: {e}")
            raise
    
    async def save_message(self, message: Dict[str, Any]) -> str:
        """
        Save a chat message to history
        
        Args:
            message: Message dictionary containing:
                - userId: User ID
                - question: User's question
                - answer: Bot's answer
                - answer_source: Source of answer (document/web_search/hybrid)
                - responseMode: Response mode used
                - confidence: Confidence score
                - timestamp: Message timestamp
                - conversationId: Optional conversation ID
                
        Returns:
            str: Document ID of saved message
        """
        try:
            # Prepare message data
            message_data = {
                'userId': message.get('userId'),
                'question': message.get('question'),
                'answer': message.get('answer'),
                'answer_source': message.get('answer_source', 'document'),
                'responseMode': message.get('responseMode', 'hybrid'),
                'sources': message.get('sources', []),
                'web_sources': message.get('web_sources', []),
                'confidence': message.get('confidence', 0.0),
                'responseTime': message.get('responseTime', 0),
                'userProfile': message.get('userProfile', 'unknown'),
                'timestamp': message.get('timestamp', datetime.now().isoformat()),
                'conversationId': message.get('conversationId'),
                'ttl': (datetime.now() + timedelta(days=self.message_retention_days)).isoformat()
            }
            
            # Save to Firestore
            user_id = message_data['userId']
            doc_ref = self.db.collection(self.history_collection).document(user_id).collection('messages').add(
                message_data
            )
            
            logger.info(f"Message saved for user {user_id} with ID {doc_ref[1].id}")
            return doc_ref[1].id
        
        except Exception as e:
            logger.error(f"Error saving message: {e}")
            raise
    
    def get_user_chat_history(self, user_id: str, limit: Optional[int] = None) -> List[Dict]:
        """
        Get all chat history for user
        
        Args:
            user_id: The user's ID
            limit: Maximum number of messages to return
            
        Returns:
            list: List of messages ordered by timestamp (newest first)
        """
        try:
            query = self.db.collection(self.history_collection).document(user_id).collection('messages')
            
            # Order by timestamp descending
            query = query.order_by('timestamp', direction=firestore.Query.DESCENDING)
            
            if limit:
                query = query.limit(limit)
            
            docs = query.stream()
            
            messages = []
            for doc in docs:
                msg = doc.to_dict()
                msg['id'] = doc.id
                messages.append(msg)
            
            logger.info(f"Retrieved {len(messages)} messages for user {user_id}")
            return messages
        
        except Exception as e:
            logger.error(f"Error retrieving chat history for user {user_id}: {e}")
            raise
    
    def get_recent_conversations(self, user_id: str, limit: int = 10) -> List[Dict]:
        """
        Get recent conversations (grouped by conversation ID)
        
        Args:
            user_id: The user's ID
            limit: Maximum number of conversations
            
        Returns:
            list: List of recent conversation messages
        """
        try:
            messages = self.get_user_chat_history(user_id, limit=limit * 10)
            
            # Group by conversation ID
            conversations = {}
            for msg in messages:
                conv_id = msg.get('conversationId', 'default')
                if conv_id not in conversations:
                    conversations[conv_id] = []
                conversations[conv_id].append(msg)
            
            # Return most recent conversations
            recent = list(conversations.values())[:limit]
            logger.info(f"Retrieved {len(recent)} recent conversations for user {user_id}")
            return recent
        
        except Exception as e:
            logger.error(f"Error retrieving recent conversations for user {user_id}: {e}")
            raise
    
    def get_conversation(self, conversation_id: str, user_id: str) -> Optional[Dict]:
        """
        Get specific conversation by ID
        
        Args:
            conversation_id: The conversation ID
            user_id: The user's ID (for authorization)
            
        Returns:
            dict: Conversation data or None if not found
        """
        try:
            messages = self.get_user_chat_history(user_id)
            
            # Filter by conversation ID
            conversation_messages = [
                msg for msg in messages 
                if msg.get('conversationId') == conversation_id
            ]
            
            if not conversation_messages:
                logger.warning(f"Conversation {conversation_id} not found for user {user_id}")
                return None
            
            return {
                'id': conversation_id,
                'userId': user_id,
                'messages': conversation_messages,
                'messageCount': len(conversation_messages),
                'startTime': conversation_messages[-1].get('timestamp'),  # First message
                'lastUpdate': conversation_messages[0].get('timestamp'),  # Last message
            }
        
        except Exception as e:
            logger.error(f"Error retrieving conversation {conversation_id}: {e}")
            raise
    
    def delete_user_chat_history(self, user_id: str) -> int:
        """
        Delete all chat history for user
        
        Args:
            user_id: The user's ID
            
        Returns:
            int: Number of messages deleted
        """
        try:
            messages_ref = self.db.collection(self.history_collection).document(user_id).collection('messages')
            docs = messages_ref.stream()
            
            count = 0
            batch = self.db.batch()
            
            for doc in docs:
                batch.delete(doc.reference)
                count += 1
                
                # Commit in batches of 500 (Firestore limit)
                if count % 500 == 0:
                    batch.commit()
                    batch = self.db.batch()
            
            # Final commit
            if count % 500 != 0:
                batch.commit()
            
            logger.info(f"Deleted {count} messages for user {user_id}")
            return count
        
        except Exception as e:
            logger.error(f"Error deleting chat history for user {user_id}: {e}")
            raise
    
    def delete_message(self, user_id: str, message_id: str) -> bool:
        """
        Delete a specific message
        
        Args:
            user_id: The user's ID
            message_id: The message ID to delete
            
        Returns:
            bool: True if successful
        """
        try:
            self.db.collection(self.history_collection).document(user_id).collection('messages').document(
                message_id
            ).delete()
            
            logger.info(f"Deleted message {message_id} for user {user_id}")
            return True
        
        except Exception as e:
            logger.error(f"Error deleting message {message_id}: {e}")
            raise
    
    def search_conversations(self, user_id: str, query_text: str) -> List[Dict]:
        """
        Search through user's chat history
        
        Args:
            user_id: The user's ID
            query_text: Search query text
            
        Returns:
            list: Messages matching search query
        """
        try:
            messages = self.get_user_chat_history(user_id, limit=1000)
            
            query_lower = query_text.lower()
            results = []
            
            for msg in messages:
                # Search in question and answer
                if (query_lower in msg.get('question', '').lower() or 
                    query_lower in msg.get('answer', '').lower()):
                    results.append(msg)
            
            logger.info(f"Found {len(results)} messages matching '{query_text}' for user {user_id}")
            return results
        
        except Exception as e:
            logger.error(f"Error searching conversations for user {user_id}: {e}")
            raise
    
    def get_chat_statistics(self, user_id: str) -> Dict[str, Any]:
        """
        Get chat statistics for user
        
        Args:
            user_id: The user's ID
            
        Returns:
            dict: Statistics including message count, avg response time, top topics
        """
        try:
            messages = self.get_user_chat_history(user_id, limit=1000)
            
            if not messages:
                return {
                    'total_messages': 0,
                    'total_conversations': 0,
                    'avg_response_time': 0,
                    'last_activity': None,
                    'joined_date': None,
                    'top_topics': [],
                    'average_confidence': 0,
                    'message_sources': {}
                }
            
            # Calculate statistics
            total_messages = len(messages)
            avg_response_time = sum(m.get('responseTime', 0) for m in messages) / total_messages if messages else 0
            total_conversations = len(set(m.get('conversationId') for m in messages))
            
            # Get source distribution
            source_counts = Counter(m.get('answer_source', 'unknown') for m in messages)
            sources = {source: count for source, count in source_counts.items()}
            
            # Extract top topics from questions
            top_topics = self._extract_top_topics([m.get('question', '') for m in messages], limit=5)
            
            # Get average confidence
            avg_confidence = sum(m.get('confidence', 0) for m in messages) / total_messages if messages else 0
            
            stats = {
                'total_messages': total_messages,
                'total_conversations': total_conversations,
                'avg_response_time': round(avg_response_time, 2),
                'avg_confidence': round(avg_confidence, 2),
                'last_activity': messages[0].get('timestamp'),  # Newest message
                'joined_date': messages[-1].get('timestamp'),   # Oldest message
                'top_topics': top_topics,
                'message_sources': sources
            }
            
            logger.info(f"Generated statistics for user {user_id}: {total_messages} messages")
            return stats
        
        except Exception as e:
            logger.error(f"Error calculating statistics for user {user_id}: {e}")
            raise
    
    def _extract_top_topics(self, questions: List[str], limit: int = 5) -> List[str]:
        """
        Extract most common topics/keywords from questions
        
        Args:
            questions: List of question strings
            limit: Maximum number of topics to return
            
        Returns:
            list: Top topics/keywords
        """
        stop_words = {
            'what', 'when', 'where', 'how', 'why', 'is', 'the', 'a', 'an', 
            'and', 'or', 'in', 'on', 'at', 'to', 'for', 'of', 'with', 'by'
        }
        
        keywords = []
        for question in questions:
            words = question.lower().split()
            # Filter out stop words and short words
            filtered = [w for w in words if w not in stop_words and len(w) > 3]
            keywords.extend(filtered)
        
        if not keywords:
            return []
        
        # Get most common keywords
        common = Counter(keywords).most_common(limit)
        return [word for word, count in common]
    
    def export_chat_history(self, user_id: str) -> Dict[str, Any]:
        """
        Export complete chat history for user
        
        Args:
            user_id: The user's ID
            
        Returns:
            dict: Exportable chat history data
        """
        try:
            messages = self.get_user_chat_history(user_id)
            
            export_data = {
                'userId': user_id,
                'exportDate': datetime.now().isoformat(),
                'totalMessages': len(messages),
                'messages': messages
            }
            
            logger.info(f"Prepared export of {len(messages)} messages for user {user_id}")
            return export_data
        
        except Exception as e:
            logger.error(f"Error exporting chat history for user {user_id}: {e}")
            raise
    
    def cleanup_expired_messages(self) -> int:
        """
        Clean up messages past retention period (admin operation)
        
        Returns:
            int: Number of messages deleted
        """
        try:
            expiration_date = datetime.now() - timedelta(days=self.message_retention_days)
            count = 0
            
            # Note: This requires a scheduled task
            logger.info(f"Cleanup would remove messages before {expiration_date}")
            return count
        
        except Exception as e:
            logger.error(f"Error in cleanup operation: {e}")
            raise

File: backend/functions/services/chatHistoryService.js
/**
 * backend/functions/services/chatHistoryService.js
 * Chat history management service for storing and retrieving conversations
 */

const admin = require("firebase-admin");
const logger = require("../utils/logger");

class ChatHistoryService {
  constructor() {
    this.db = admin.firestore();
    this.historyCollection = "chat_history";
    this.messageRetentionDays = 90;
    logger.info("ChatHistoryService initialized");
  }

  /**
   * Save a chat message to history
   * @param {Object} message - Message data containing userId, question, answer, etc.
   * @returns {Promise<string>} Document ID of saved message
   */
  async saveMessage(message) {
    try {
      // Prepare message data
      const messageData = {
        userId: message.userId,
        question: message.question,
        answer: message.answer,
        answer_source: message.answer_source || "document",
        responseMode: message.responseMode || "hybrid",
        sources: message.sources || [],
        web_sources: message.web_sources || [],
        confidence: message.confidence || 0.0,
        responseTime: message.responseTime || 0,
        userProfile: message.userProfile || "unknown",
        timestamp: message.timestamp || new Date().toISOString(),
        conversationId: message.conversationId,
        ttl: new Date(
          Date.now() + this.messageRetentionDays * 24 * 60 * 60 * 1000,
        ).toISOString(),
      };

      // Save to Firestore
      const userId = messageData.userId;
      const docRef = await this.db
        .collection(this.historyCollection)
        .doc(userId)
        .collection("messages")
        .add(messageData);

      logger.info(`Message saved for user ${userId} with ID ${docRef.id}`);
      return docRef.id;
    } catch (error) {
      logger.error(`Error saving message: ${error.message}`);
      throw error;
    }
  }

  /**
   * Get all chat history for user
   * @param {string} userId - The user's ID
   * @param {number} limit - Maximum number of messages to return
   * @returns {Promise<Array>} List of messages ordered by timestamp (newest first)
   */
  async getUserChatHistory(userId, limit = null) {
    try {
      let query = this.db
        .collection(this.historyCollection)
        .doc(userId)
        .collection("messages")
        .orderBy("timestamp", "desc");

      if (limit) {
        query = query.limit(limit);
      }

      const snapshot = await query.get();
      const messages = [];

      snapshot.forEach((doc) => {
        const msg = doc.data();
        msg.id = doc.id;
        messages.push(msg);
      });

      logger.info(`Retrieved ${messages.length} messages for user ${userId}`);
      return messages;
    } catch (error) {
      logger.error(
        `Error retrieving chat history for user ${userId}: ${error.message}`,
      );
      throw error;
    }
  }

  /**
   * Get recent conversations (grouped by conversation ID)
   * @param {string} userId - The user's ID
   * @param {number} limit - Maximum number of conversations
   * @returns {Promise<Array>} List of recent conversation messages
   */
  async getRecentConversations(userId, limit = 10) {
    try {
      const messages = await this.getUserChatHistory(userId, limit * 10);

      // Group by conversation ID
      const conversations = {};
      messages.forEach((msg) => {
        const convId = msg.conversationId || "default";
        if (!conversations[convId]) {
          conversations[convId] = [];
        }
        conversations[convId].push(msg);
      });

      // Return most recent conversations
      const recent = Object.values(conversations).slice(0, limit);
      logger.info(
        `Retrieved ${recent.length} recent conversations for user ${userId}`,
      );
      return recent;
    } catch (error) {
      logger.error(
        `Error retrieving recent conversations for user ${userId}: ${error.message}`,
      );
      throw error;
    }
  }

  /**
   * Get specific conversation by ID
   * @param {string} conversationId - The conversation ID
   * @param {string} userId - The user's ID (for authorization)
   * @returns {Promise<Object|null>} Conversation data or null if not found
   */
  async getConversation(conversationId, userId) {
    try {
      const messages = await this.getUserChatHistory(userId);

      // Filter by conversation ID
      const conversationMessages = messages.filter(
        (msg) => msg.conversationId === conversationId,
      );

      if (conversationMessages.length === 0) {
        logger.warning(
          `Conversation ${conversationId} not found for user ${userId}`,
        );
        return null;
      }

      return {
        id: conversationId,
        userId,
        messages: conversationMessages,
        messageCount: conversationMessages.length,
        startTime:
          conversationMessages[conversationMessages.length - 1].timestamp,
        lastUpdate: conversationMessages[0].timestamp,
      };
    } catch (error) {
      logger.error(
        `Error retrieving conversation ${conversationId}: ${error.message}`,
      );
      throw error;
    }
  }

  /**
   * Delete all chat history for user
   * @param {string} userId - The user's ID
   * @returns {Promise<number>} Number of messages deleted
   */
  async deleteUserChatHistory(userId) {
    try {
      const messagesRef = this.db
        .collection(this.historyCollection)
        .doc(userId)
        .collection("messages");

      const snapshot = await messagesRef.get();
      let count = 0;
      const batch = this.db.batch();

      snapshot.forEach((doc) => {
        batch.delete(doc.ref);
        count++;

        // Commit in batches of 500 (Firestore limit)
        if (count % 500 === 0) {
          batch.commit();
        }
      });

      // Final commit
      if (count % 500 !== 0) {
        await batch.commit();
      }

      logger.info(`Deleted ${count} messages for user ${userId}`);
      return count;
    } catch (error) {
      logger.error(
        `Error deleting chat history for user ${userId}: ${error.message}`,
      );
      throw error;
    }
  }

  /**
   * Delete a specific message
   * @param {string} userId - The user's ID
   * @param {string} messageId - The message ID to delete
   * @returns {Promise<boolean>} True if successful
   */
  async deleteMessage(userId, messageId) {
    try {
      await this.db
        .collection(this.historyCollection)
        .doc(userId)
        .collection("messages")
        .doc(messageId)
        .delete();

      logger.info(`Deleted message ${messageId} for user ${userId}`);
      return true;
    } catch (error) {
      logger.error(`Error deleting message ${messageId}: ${error.message}`);
      throw error;
    }
  }

  /**
   * Search through user's chat history
   * @param {string} userId - The user's ID
   * @param {string} queryText - Search query text
   * @returns {Promise<Array>} Messages matching search query
   */
  async searchConversations(userId, queryText) {
    try {
      const messages = await this.getUserChatHistory(userId, 1000);

      const queryLower = queryText.toLowerCase();
      const results = messages.filter(
        (msg) =>
          queryLower.includes(msg.question?.toLowerCase() || "") ||
          queryLower.includes(msg.answer?.toLowerCase() || ""),
      );

      logger.info(
        `Found ${results.length} messages matching '${queryText}' for user ${userId}`,
      );
      return results;
    } catch (error) {
      logger.error(
        `Error searching conversations for user ${userId}: ${error.message}`,
      );
      throw error;
    }
  }

  /**
   * Get chat statistics for user
   * @param {string} userId - The user's ID
   * @returns {Promise<Object>} Statistics including message count, avg response time, top topics
   */
  async getChatStatistics(userId) {
    try {
      const messages = await this.getUserChatHistory(userId, 1000);

      if (messages.length === 0) {
        return {
          total_messages: 0,
          total_conversations: 0,
          avg_response_time: 0,
          avg_confidence: 0,
          last_activity: null,
          joined_date: null,
          top_topics: [],
          message_sources: {},
        };
      }

      // Calculate statistics
      const totalMessages = messages.length;
      const avgResponseTime =
        messages.reduce((sum, m) => sum + (m.responseTime || 0), 0) /
        totalMessages;

      const conversationIds = new Set(messages.map((m) => m.conversationId));
      const totalConversations = conversationIds.size;

      // Get source distribution
      const sourceCounts = {};
      messages.forEach((m) => {
        const source = m.answer_source || "unknown";
        sourceCounts[source] = (sourceCounts[source] || 0) + 1;
      });

      // Extract top topics
      const topTopics = this._extractTopTopics(
        messages.map((m) => m.question || ""),
        5,
      );

      // Get average confidence
      const avgConfidence =
        messages.reduce((sum, m) => sum + (m.confidence || 0), 0) /
        totalMessages;

      const stats = {
        total_messages: totalMessages,
        total_conversations: totalConversations,
        avg_response_time: Math.round(avgResponseTime * 100) / 100,
        avg_confidence: Math.round(avgConfidence * 100) / 100,
        last_activity: messages[0].timestamp,
        joined_date: messages[messages.length - 1].timestamp,
        top_topics: topTopics,
        message_sources: sourceCounts,
      };

      logger.info(
        `Generated statistics for user ${userId}: ${totalMessages} messages`,
      );
      return stats;
    } catch (error) {
      logger.error(
        `Error calculating statistics for user ${userId}: ${error.message}`,
      );
      throw error;
    }
  }

  /**
   * Extract most common topics/keywords from questions
   * @private
   * @param {Array<string>} questions - List of question strings
   * @param {number} limit - Maximum number of topics to return
   * @returns {Array<string>} Top topics/keywords
   */
  _extractTopTopics(questions, limit = 5) {
    const stopWords = new Set([
      "what",
      "when",
      "where",
      "how",
      "why",
      "is",
      "the",
      "a",
      "an",
      "and",
      "or",
      "in",
      "on",
      "at",
      "to",
      "for",
      "of",
      "with",
      "by",
    ]);

    const keywords = [];
    questions.forEach((question) => {
      const words = question.toLowerCase().split(/\s+/);
      words.forEach((w) => {
        if (!stopWords.has(w) && w.length > 3) {
          keywords.push(w);
        }
      });
    });

    if (keywords.length === 0) return [];

    // Count occurrences
    const counts = {};
    keywords.forEach((k) => {
      counts[k] = (counts[k] || 0) + 1;
    });

    // Get most common
    return Object.entries(counts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, limit)
      .map(([word]) => word);
  }

  /**
   * Export complete chat history for user
   * @param {string} userId - The user's ID
   * @returns {Promise<Object>} Exportable chat history data
   */
  async exportChatHistory(userId) {
    try {
      const messages = await this.getUserChatHistory(userId);

      const exportData = {
        userId,
        exportDate: new Date().toISOString(),
        totalMessages: messages.length,
        messages,
      };

      logger.info(
        `Prepared export of ${messages.length} messages for user ${userId}`,
      );
      return exportData;
    } catch (error) {
      logger.error(
        `Error exporting chat history for user ${userId}: ${error.message}`,
      );
      throw error;
    }
  }
}

module.exports = ChatHistoryService;
